Client/SafeBrowsing/GSafeBrowsingDatabase.swift

```diff
@@ -0,0 +1,509 @@
+                //Hashes must be sorted
+                //hashes.sort(by: { $0.lexicographicallyPrecedes($1) })
+                
+                response.removals.forEach({
+                    $0.rawIndices?.indices.forEach({
+                        if Int($0) > 0 && Int($0) < hashes.count {
+                            hashes[Int($0)] = Data()
+                        }
+                    })
+                })
+                
+                hashes = hashes.filter({ !$0.isEmpty })
+                
+                response.additions.forEach({
+                    guard let rawHashes = $0.rawHashes else {
+                        return
+                    }
+                    
+                    if let data = Data(base64Encoded: rawHashes.rawHashes) {
+                        let strideSize = Int(rawHashes.prefixSize)
+                        
+                        for i in stride(from: 0, to: data.count, by: strideSize) {
+                            let startIndex = data.index(data.startIndex, offsetBy: i)
+                            let endIndex = data.index(startIndex, offsetBy: strideSize)
+                            hashes.append(data.subdata(in: startIndex..<endIndex))
+                        }
+                    }
+                })
+                
+                //Hashes must be sorted
+                hashes.sort(by: { $0.lexicographicallyPrecedes($1) })
+                
+                threat.hashes?.forEach({
+                    if let hash = $0 as? ThreatHash {
+                        backgroundContext.delete(hash)
+                    }
+                })
+                
+                threat.hashes = NSOrderedSet(array: hashes.map({
+                    let hash = ThreatHash(context: backgroundContext)
+                    hash.hashPrefix = $0.rawString()
+                    hash.hashData = $0
+                    return hash
+                }))
+                
+                if !response.checksum.sha256.isEmpty {
+                    threat.checksum = response.checksum.sha256
+                }
+                
+                if !self.validate(response.checksum.sha256, hashes) {
+                    backgroundContext.rollback()
+                    return completion(SafeBrowsingError("Threat List Checksum Mismatch"))
+                }
+                
+                self.saveContext(backgroundContext)
+            }
+        })
+    }
+    
+    func scheduleUpdate(onNeedsUpdating: @escaping () -> Void) {
+        updateTimer?.invalidate()
+        updateTimer = Timer.scheduledTimer(withTimeInterval: self.lastFetchWaitDuration, repeats: false) {
+            $0.invalidate()
+            DispatchQueue.main.async {
+                onNeedsUpdating()
+            }
+        }
+    }
+    
+    private func calculateBackoffTime(_ numberOfRetries: Int16) -> Double {
+        let minutes = Double(1 << Int(numberOfRetries - 1)) * (15.0 * 60.0) * Double.random(in: 0...1) + 1
```

Comment:
If the unit is "minutes" here, are we really supposed to multiply `15` by `60`?

Reply:
When dealing with time, I personally like to stick the unit in the function name because it's so easy to forget that some functions take milliseconds and others take seconds (or minutes in this case).

Reply:
Here's another subtle difference between the above code and the spec:

- the above code does `15 * RAND + 1` which corresponds to `(15 * RAND) + 1`
- the spec says [`15 * (RAND + 1)`](https://developers.google.com/safe-browsing/v4/request-frequency#back-off-mode)

Reply:
Fixed.